_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,387 @@
+
+```scala
+import Tup._
+
+extension [A, T <: Tup] (a: A) def *: (t: T): TCons[A, T] =
+  TCons(a, t)
+```
+So that we can write:
+
+```scala
+1 *: "2" *: EmpT
+```
+
+## Concatenating tuples
+
+Now let's focus on `concat`, which could look like this:
+```scala
+import Tup._
+
+def concat[L <: Tup, R <: Tup](left: L, right: R): Tup =
+  left match
+    case EmpT => right
+    case TCons(head, tail) => TCons(head, concat(tail, right))
+```
+
+Let's analyze the algorithm line by line:
+`L` and `R` are the type of the left and right tuple. We require
+them to be a subtype of `Tup` because we want to concatenate tuples.
+Why not using `Tup` directly? Because in this way we receive more specific
+information about the two arguments.
+Then we proceed recursively by case: if the left tuple is empty,
+the result of the concatenation is just the right tuple.
+Otherwise the result is the current head followed by the result of
+concatenating the tail with the other tuple.
+
+If we test the function, it seems to work:
+```scala
+val left = 1 *: 2 *: EmpT
+val right = 3 *: 4 *: EmpT
+
+concat(left, right) // TCons(1,TCons(2,TCons(3, TCons(4,EmpT))))
+```
+
+So everything seems good. However we can have more safety.
+For instance the following code is perfectly fine:
+```scala
+def concat[L <: Tup, R <: Tup](left: L, right: R): Tup = left
+```
+Because the returned type is just a tuple, we do not check anything else.
+This means that the function can return an arbitrary tuple,
+the compiler cannot check that returned value consists of the concatenation
+of the two tuples. In other words, we need a type to indicate that
+the return of this function is all the types of `left` followed
+by all the types of the elements of `right`.
+
+Can we make it so that the compiler verifies that we are indeed
+returning a tuple consisting of the correct elements ?
+
+In Scala 3 it is now possible, without requiring external libraries!
+
+## A new type for the result of `concat`
+
+We know that we need to focus on the return type. We can define this the return
+type exactly as we have just described it.
+Let's call this type `Concat` to mirror the name of the function.
+
+```scala
+type Concat[L <: Tup, R <: Tup] <: Tup = L match
+  case EmpT.type => R
+  case TCons[h, t] => TCons[h, Concat[t, R]]
+```
+
+You can see that the implementation closely follows the one
+above for the method.
+To use it we need to massage a bit the method implementation and
+to change its return type:
+
+```scala
+def concat[L <: Tup, R <: Tup](left: L, right: R): Concat[L, R] =
+  left match
+    case _: EmpT.type => right
+    case cons: TCons[head, tail] => TCons(cons.head, concat(cons.tail, right))
+```
+
+We use here a combination of match types and a form of dependent types called
+*dependently match types*. There are some quirks to it as you might have noticed:
```

Comment:
isn't it just "dependent match types"? (I don't actually know — but “dependently match types” isn't correct English)